### PR TITLE
aws-sigv4: fix canonical path encoding

### PR DESCRIPTION
**Pull request by outscale-mgo**
fix: #13754

I refactor canon_query, so I could use the encoding part of the function to use it in the path.

As the path doesn't encode '/', but encode '=', I had to add some conditions to know If I was doing the query or path encoding.

Also, instead of adding a `bool in_path` variable, I use `bool *found_equals` to know if the function was called for the query or path, as found_equals is used only in query_encoding.

Note: When I try replacing manually `=` with `%3D`, curl without this patch worked, so they might still be a bug out-there.

### CODE REVIEW

File: lib/http_aws_sigv4.c [hunk 1]
```diff
@@ -423,6 +423,77 @@ static int compare_func(const void *a, const void *b)
...
+ * and mark that this function is called to compute the path,
+ * if found_equals is NULL.
+ */
+static CURLcode canon_string(const char *q, size_t str_len,
+                             struct dynbuf *dq, bool *found_equals)
+{
+  CURLcode result = CURLE_OK;
+  size_t len;
+
+  for(len = str_len; len && !result; q++, len--) {
```
Comment (bagder) [inline lib/http_aws_sigv4.c:438]: You can remove a variable like this:

```suggestion
static CURLcode canon_string(const char *q, size_t len,
                             struct dynbuf *dq, bool *found_equals)
{
  CURLcode result = CURLE_OK;

  for(; len && !result; q++, len--) {
```

File: lib/http_aws_sigv4.c [hunk 2]
```diff
@@ -460,54 +531,11 @@ static CURLcode canon_query(struct Curl_easy *data,
...
-          char out[3]={'%'};
-          out[1] = hex[((unsigned char)*q)>>4];
-          out[2] = hex[*q & 0xf];
-          result = Curl_dyn_addn(dq, out, 3);
-          break;
-        }
-        }
-      }
-    }
+    canon_string(q, ap->len, dq, &found_equals);
```
Comment (bagder) [inline lib/http_aws_sigv4.c:538]: this should receive the return code in `result` should it not?

```suggestion
    result = canon_string(q, ap->len, dq, &found_equals);
```

**Comment by outscale-mgo** [comment]
done, thanks for the review.

**Review by bagder — APPROVED** [review]

**Comment by bagder** [comment]
Thanks!

**Pull request closed without merging**